### ISSUE TITLE
[Code Addition Request]: PassClass: AI-Powered Password Strength Classifier

### ISSUE THREAD

**Issue by SK8-infi** [UTSAVS26/PyVerse#1622]
### Have you completed your first issue?

- [x] I have completed my first issue

### Guidelines

- [x] I have read the guidelines
- [x] I have the link to my latest merged PR

### Latest Merged PR Link

https://github.com/UTSAVS26/PyVerse/pull/1567

### 📌 Project Overview

**PassClass** is an intelligent password strength evaluation system that:

1. **Generates synthetic passwords** with varying complexity,
2. **Labels them automatically** (weak / medium / strong) using rule-based heuristics,
3. **Trains an ML or NLP model** to predict password strength just from the text.

No dataset required — everything is self-generated.

---

### 💡 Use Cases

* 🛡️ Password strength meters for login/signup pages
* 🔐 Training AI to understand secure vs insecure credentials
* 🧠 NLP-based classification using synthetically labeled data

---

### 📂 Folder Structure

```
passclass/
│
├── data/
│   ├── generate_passwords.py       # Rule-based password generator
│   └── password_dataset.csv        # Auto-labeled data
│
├── labeling/
│   └── labeler.py                  # Heuristic rules to label strength
│
├── models/
│   ├── tfidf_classifier.py         # Classic ML: SVM / Random Forest on TF-IDF
│   └── transformer_classifier.py   # Optional: BERT-based classifier
│
├── training/
│   ├── train_model.py              # Train and validate model
│   └── evaluate.py                 # Accuracy, precision, F1, etc.
│
├── ui/
│   └── streamlit_app.py            # Interactive web app for testing strength
│
├── utils/
│   └── metrics.py                  # Scoring and visualization
│
├── tests/
│   └── test_generation.py
│
├── requirements.txt
├── README.md
└── LICENSE
```

---

#### ✅ Synthetic Password Generator:

* Weak: `"pass123"`, `"abc"`, `"hello123"`
* Medium: `"Hello2023!"`, `"book@789"`
* Strong: `"G7^s9L!zB1m"`, `"tR#8$!XmPq@"`

#### 🧪 Labeling Heuristics:

* Length score
* Presence of uppercase, lowercase, digit, special characters
* Common patterns/words blacklist

Example rule:

```python
if length < 6 or only letters: label = "weak"
elif has_digit and special and length > 10: label = "strong"
else: label = "medium"
```

---

#### 🔹 ML Models:

* TF-IDF + Logistic Regression / Random Forest
* Character n-gram features

#### 🔸 NLP Models:

* BERT-like model fine-tuned for classification
* Input: password as a string → Output: strength category

---

### 🧪 Sample Evaluation Output

```
Training Accuracy: 94.2%  
Validation F1 Score: 91.7%  
Strong passwords detected: ✅  
False positives (weak marked strong): 2.8%
```

---

### 🔧 Features

* 🔐 100% synthetic password dataset (no external data)
* ⚙️ Configurable password rules
* 📊 Strength prediction with confidence score
* 🌐 Streamlit web app demo (enter password → see prediction)
* 📈 Model performance charts

---

### 📘 README Highlights

**How it works:**

* Generate passwords
* Label them using security heuristics
* Train a classifier to learn the underlying patterns of secure vs insecure passwords

**Dependencies:**

* Python
* Scikit-learn, Pandas
* HuggingFace Transformers (optional)
* Streamlit (for UI)

---

### 🧪 Possible Enhancements

* Add entropy and zxcvbn-style complexity scoring
* Train on leaked/common passwords (like RockYou) for realism
* Export model as a browser-usable `onnx` or `tf.js` app
* Suggest stronger versions of weak inputs

---

### Full Name

Shivansh Katiyar

### Participant Role

SSOC
